### PR TITLE
Business logic and tests

### LINKED ISSUE #13: ViewModel Tests Failing
**Describe the bug**
ViewModel tests pass in isolation but fail in tandem.

**To Reproduce**
1. Go to com/example/tictactoe/model/TicTacToeViewModelTest.kt
2. Click on each individual test: All pass
3. Scroll up to top and run entire class: 
These tests pass:
- set_top_left_to_bottom_right_diagonal_test_diagonal_returns_true()
- set_icon_as_x_confirm_returns_x_icon()
- set_icon_as_illegal_arg_confirm_returns_exception_and_error_icon()
The rest all fail.
4. Run any in isolation and they pass again

**Expected behavior**
With the @Before and @After blocks I set, I expected each test to run as if on a blank slate. However clearly, some state is passing over. It is unclear why only one of the many identical 3-in-a-row-returns-true tests passes. And of them all, not the first test in the file.

### PR DESCRIPTION
**Pull request by Jherscu**
# Description

Create the business logic for the working of the game, as well as the interface to reflect changes in the UI.

Includes tests in multiple scopes proving that the code works both independently and in integration.

- [x] New feature (non-breaking change which adds functionality)

Closes #13

# How Has This Been Tested?

## GameFragmentTest:
- [x] confirms updation of content description
## NavigationTest: 
- [x] confirms that navigation works as expected
## TicTacToeViewModelTest:
- [x] set_EXAMPLE_row_test_row_returns_true() suite: confirms that getting three in a row returns a win
- [x] set_top_left_to_bottom_right_diagonal_test_live_data_values_change(): confirms isWin and winningPlayer liveData 
       update properly on win
- [x] set_icon_as_x_confirm_returns_x_icon(): confirms viewModel.getIcon() functionality
- [x] set_icon_as_illegal_arg_confirm_returns_exception_and_error_icon(): confirms viewModel.getIcon() edge case

**Test Configuration**:
* Firmware version: API 31
* Hardware: Pixel 4
* Toolchain: Tests launched with Gradle on emulator
* SDK: Android SDK

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by j-roskopf — APPROVED** [review]
it looks solid thus far! you might want to consider using something like a class or view model instead of an object. an object has the property in kotlin that it is a singleton, so there is only ever one instance of `GameBoardModel` throughout the application. this will mostly be fine, but something to consider around testing is you wouldn't want the statefulness of `GameBoardModel` to leak through between tests.

**Comment by Jherscu** [comment]
@j-roskopf Gotcha! I actually had the same thought after posting this. And refactored it to have an underlying interface which I made a fake of for testing. I want there to only be one instance throughout the app that I can reset after each game.
 
If I hadn't have done that, would an @After notation with some cleanup have taken care of it in testing, or would it just be preferable in general to make it a class?

I wasn't sure if it could then create multiple conflicting instances when called from different modules. It gets referenced a lot.

Will push the aforementioned code shortly.

**Comment by j-roskopf** [comment]
using an interface and then creating a 'real' and a 'fake' implementation for prod / testing is a great strategy. 

using @after in your unit tests to cleanup the state would have also solved it yeah

File: app/src/main/java/com/example/tictactoe/ui/GameFragment.kt [hunk 1]
```diff
@@ -53,10 +53,220 @@ class GameFragment : Fragment() {
...
+            gridUpperCenter.setOnClickListener {
+
+                // If the box has already been populated with a value
+                // return a warning toast
+                viewModel.clickBox(0, 1).toastWhenFalse()
+                // Observe state change and update views
+                observeGameBoard()
+
+                // Checks game board for win
+                viewModel.isThreeInARow()
```
Comment (j-roskopf) [inline app/src/main/java/com/example/tictactoe/ui/GameFragment.kt]: could `clickBox` just call `isThreeInARow` so we don't have to do it every time?

File: gradle.properties [hunk 2]
```diff
@@ -20,4 +20,6 @@ kotlin.code.style=official
 # Enables namespacing of each library's R class so that its R class includes only the
 # resources declared in the library itself and none from the library's dependencies,
 # thereby reducing the size of the R class for that library
-android.nonTransitiveRClass=true
\ No newline at end of file
+android.nonTransitiveRClass=true
+# migrate old dependencies automatically to androidX equivalents
+android.enableJetifier=true
```
Comment (j-roskopf) [inline gradle.properties]: is this needed? i didn't see any older support dependencies in the project

**Review by j-roskopf — APPROVED** [review]
overall I think things are looking really good! i really loved the abstractions created with the interface and the attention to testing. those are definitely some higher level concepts you seem to have gotten early on which is awesome! it definitely seems like you have a real aptitude for picking up good coding practices. stick with it!

File: gradle.properties (hunk 2 above)
Comment (Jherscu) [reply gradle.properties]: After moving to a higher version of Gradle I kept getting notices that something was outdated, but I couldn't track down what it was exactly. So I figured jetifier would handle it. The error is gone but I still am a little miffed I couldn't track down what it fixed.

File: app/src/main/java/com/example/tictactoe/ui/GameFragment.kt (hunk 1 above)
Comment (Jherscu) [reply app/src/main/java/com/example/tictactoe/ui/GameFragment.kt]: I'll take a look! Thanks

File: gradle.properties (hunk 2 above)
Comment (Jherscu) [reply gradle.properties]: Found it! Thanks for the prodding to look into it further

File: app/src/test/java/com/example/tictactoe/model/TicTacToeViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+        FakeGameBoardModel.testWin(2, FakeGameBoardModel.Vertical())
+
+        assertThat("Testing win did not return true for right column",
+            viewModel.testRow(FakeGameBoardPart.COLUMN_RIGHT.part, FakeGameBoardPart.BOX_BOTTOM_RIGHT.part),
+            `is`(true)
+        )
+    }
+
+    @Test
```
Comment (Jherscu) [inline app/src/test/java/com/example/tictactoe/model/TicTacToeViewModelTest.kt:100]: For some reason when I run all the tests together this one will pass (along with the last two) but the rest fail.

Independently they all pass.

@j-roskopf Do you have any clue what I might be missing?

**Comment by Jherscu** [comment]
@j-roskopf I had multiple issues crop up around testing the singleton, as you predicted!

It makes me a little uncomfortable, but for the moment I have the fake modeled as a class. This way each test can have its own state. I tried my utmost to reset the gameboard between tests even while using the same object, but it clearly got me nowhere. I suppose I just don't understand what exactly is going on behind the scenes yet.

I don't think testing it as a class would have any catastrophic effect on the actual function of the game.

I believe that some difference in the way the tests ran compared to the actual application was causing it not to reset properly.

Resetting the game board after each game should still function normally.

If this still seems fine, I am ready to merge this branch and finalize the full functionality in a new one.

Hope you're doing well,

Thanks again for all the help!

**Comment by Jherscu** [comment]
@j-roskopf I've actually finished the full function branch as well and I'll make a pr for you to review as soon as you ok this one! If you find anything I'll work on that first of course!

File: app/build.gradle [hunk 4]
```diff
@@ -32,12 +45,12 @@ android {
     }
 
     compileOptions {
-        sourceCompatibility JavaVersion.VERSION_1_8
-        targetCompatibility JavaVersion.VERSION_1_8
+        sourceCompatibility JavaVersion.VERSION_11
```
Comment (j-roskopf) [inline app/build.gradle:48]: java 11! nice :)

File: app/build.gradle [hunk 5]
```diff
@@ -48,19 +61,24 @@ dependencies {
     implementation "androidx.appcompat:appcompat:$appcompat_version"
     implementation "com.google.android.material:material:$material_version"
     implementation "androidx.constraintlayout:constraintlayout:$constraint_layout_version"
+    implementation 'androidx.test.ext:junit-ktx:1.1.3'
```
Comment (j-roskopf) [inline app/build.gradle]: might be nice to follow the same pattern of defining the version elsewhere

File: app/src/main/java/com/example/tictactoe/model/TicTacToeViewModel.kt [hunk 6]
```diff
@@ -32,4 +56,172 @@ class TicTacToeViewModel : ViewModel() {
...
+     */
+    fun getIcon(boardValue: String?): Int {
+        return try {
+            when (boardValue) {
+                "X" -> R.drawable.icon_x
+                "O" -> R.drawable.icon_o
+                else -> throw IllegalArgumentException("Must pass valid symbol to getIcon")
+            }
+        } catch (e: IllegalArgumentException) {
+            Log.e(TAG, e.message.toString())
```
Comment (j-roskopf) [inline app/src/main/java/com/example/tictactoe/model/TicTacToeViewModel.kt]: what you'll find a lot of people do in this scenario is just use the class name `this::class.java.simpleName` is the tag - and what you'll find popular logging libraries like Timber do you for out of the box - makes it less brittle to refactors

**Review by j-roskopf — APPROVED** [review]
looks awesome! a lot of good concepts in here (tests, using fakes, etc)

File: app/build.gradle (hunk 5 above)
Comment (Jherscu) [reply app/build.gradle]: thanks! I missed that when I updated the dependency. I accidentally used a support version, which is why I needed jetifier before.

File: app/src/main/java/com/example/tictactoe/model/TicTacToeViewModel.kt (hunk 6 above)
Comment (Jherscu) [reply app/src/main/java/com/example/tictactoe/model/TicTacToeViewModel.kt]: Thank you! I've heard of timber but hadn't gotten the chance to use it yet. I did a little research and switched it out for my own custom tag.
